### ISSUE TITLE
Building Rust apps with non-Rust build dependencies

### ISSUE THREAD

**Issue by blm768** [nix-community/dream2nix#171]
I've got a Rust app I'm trying to package that ends up needing `pkg-config` for part of the build process. I haven't managed to figure out how to inject that dependency. Here's what I've been trying:

```nix
{
  inputs = {
    dream2nix.url = "github:nix-community/dream2nix";
    src.url = "github:JoNil/elf2uf2-rs";
    src.flake = false;
  };

  outputs =
    { self, dream2nix, nixpkgs, src }:
    let
      pkgs = nixpkgs.legacyPackages.x86_64-linux;
    in
    (dream2nix.lib.makeFlakeOutputs {
      systems = [ "x86_64-linux" ];
      config.projectRoot = ./.;
      source = src;
      settings = [
        {
          builder = "crane";
          translator = "cargo-lock";
        }
      ];
      packageOverrides = {
        elf2uf2-rs = {
          add-deps = {
              buildInputs = [ pkgs.pkg-config ];
          };
        };
      };
    })
    // {
      checks.x86_64-linux.elf2uf2-rs = self.packages.x86_64-linux.elf2uf2-rs;
      defaultPackage.x86_64-linux = self.packages.x86_64-linux.elf2uf2-rs;
    };
}
```

I could just be getting the override syntax wrong, but the build still isn't able to find the `pkg-config` executable. I can probably work around this by just using `buildRustPackage` directly along with [`crateOverrides`](https://nixos.org/manual/nixpkgs/stable/#handling-external-dependencies), but I figured I'd file this here in case this is a use case that should be on the radar.

**Comment by yusdacra** [comment]
You need to specify the `add-deps` override like so: `nativeBuildInputs = old: old ++ [pkgs.pkg-config];`. That should work I think. If it doesn't you can also just try `nativeBuildInputs = [pkgs.pkg-config];`. You can use `crateOverrides` like overrides by specifying and override like `add-deps.overrideAttrs = prev: {};`.

**Comment by DavHau** [comment]
@yusdacra I tried that, but the name of the derivation that is failing is `elf2uf2-rs-deps`. It's not the `elf2uf2-rs` derivation and therefore the override might not be applied.

**Comment by DavHau** [comment]
@yusdacra The `-deps` deriavitons are created by the crane library, right? So how would we apply an override there?

**Comment by DavHau** [comment]
fixed via f13621b7d315efbfadf1f6705737fd0a59766745

> We can add logic in dream2nix to get all the overrides and also apply main package overrides to the deps-only derivation if none are specified for the deps-only derivation. But as I thought more about it I think it would be better to just add some documentation for the crane builder that tells you it will use the <pkg name>-deps for deps-only derivation. Feels like automatically doing that is too much of a hidden behaviour.

Sorry I read that too late. Should I undo the fix?

Concerning the documentation:
Maybe it's enough to have one docu page per subsystem with different chapters?
And yes, I think it would be interesting to have all of it available via cmdline tool. Just `man` would be sufficient I guess? And also a website. Maybe something like mdBook?

**Comment by blm768** [comment]
Yep; switching to `nativeBuildInputs` and adding the inputs to both derivations did the trick. Thanks!

It looks like `crane`'s docs describe this pretty well; I just hadn't gotten that far (and somehow ended up trying to read the wrong builder's source; go figure).

**Comment by yusdacra** [comment]
> fixed via [f13621b](https://github.com/nix-community/dream2nix/commit/f13621b7d315efbfadf1f6705737fd0a59766745)
> 
> > We can add logic in dream2nix to get all the overrides and also apply main package overrides to the deps-only derivation if none are specified for the deps-only derivation. But as I thought more about it I think it would be better to just add some documentation for the crane builder that tells you it will use the -deps for deps-only derivation. Feels like automatically doing that is too much of a hidden behaviour.
> 
> Sorry I read that too late. Should I undo the fix?
> 
> Concerning the documentation: Maybe it's enough to have one docu page per subsystem with different chapters? And yes, I think it would be interesting to have all of it available via cmdline tool. Just `man` would be sufficient I guess? And also a website. Maybe something like mdBook?

Yeah, I reverted the commit. I'm working on documentation, will come up with something for that so users can know about the behavior.

**Comment by yusdacra** [comment]
FYI this behaviour is now documented and it should be available in the rust section of [the book](https://nix-community.github.io/dream2nix).
